pkg/util/subnet/subnetsupdater.go

```diff
@@ -0,0 +1,74 @@
+package subnet
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+import (
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+type Updater interface {
+	AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet)
+}
+
+type UpdaterManager struct {
+}
+
+func (um *UpdaterManager) AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet) {
+	for _, subnet := range subnets {
+		subnetChanged := addEndpointsToSubnet(endpoints, subnet)
+		if subnetChanged {
+			updatedSubnets = append(updatedSubnets, subnet)
+		}
+	}
+
+	return updatedSubnets
+}
+
+func addEndpointsToSubnet(endpoints []string, subnet *mgmtnetwork.Subnet) (subnetChanged bool) {
+	for _, endpoint := range endpoints {
+		endpointFound, serviceEndpointPtr := subnetContainsEndpoint(subnet, endpoint)
+		if !endpointFound {
+			addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+			continue
+		}
+
+		if serviceEndpointPtr.ProvisioningState != mgmtnetwork.Succeeded {
+			addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+		}
```

Comment:
I don't really get this part. It does the exact same thing, and the `continue` isn't doing anything. So i don't think it has any value here

Reply:
also, if it makes no sense adding an endpoint twice in the same subnet (given that these are domain names or something I would assume it does not), it would probably be better to have an idempotent `addEndpointToSubnet` function which return true if it was added rather than this one.

Reply:
"it makes no sense adding an endpoint twice in the same subnet"
I have not decided to do it that way 🙂

I agree with you that it is a bit strange, BUT this PR is a refactor and this is how it was done before. I am not sure if changing the behavior now is a good thing because if the code does not work once merged (as this is touching infrastructure), all the work done here will be reverted. That's why I would just ensure the refactor is good and in a future pull request we could change that behavior (because thanks to the refactor will be easy to change).

In the original code, you can check "pkg/cluster/storageaccounts.go:47-48" and you will see that it sets "found to true" when it both finds the serviceEndpoint and the state of the serviceEndpoint is == mgmtnetwork.Succeeded. If the endpoint is present but the state is != "mgmtnetwork.Succeeded", found will still be false and it will add the service endpoint to the subnet even if the service endpoint is already present. And that is what my code does ( on porpuse). 

A part from that, I think that the code block you have marked in this conversation can be simplified to
```go
if !endpointFound ||
			serviceEndpointPtr != nil && serviceEndpointPtr.ProvisioningState != mgmtnetwork.Succeeded {
			addEndpointToSubnet(endpoint, subnet)
			subnetChanged = true
		}

```
so will do it now

Reply:
> "it makes no sense adding an endpoint twice in the same subnet"
> I have not decided to do it that way slightly_smiling_face

I didn't say, "it makes no sense", it was a conditional statement. 

Your function names aren't chosen properly then. Because the current ones imply that in `addEndpointsToSubnet` is just calling `addEndpointToSubnet` for multiple endpoints.

Reply:
Also I missed it when you updated it, but when is the pointer nil and endpointFound==true ? 
Your pointer check is superfluous

Reply:
I agree on that, I will improve it with a Go comment now to clarify it.
 
Also, how should I name the functions?
"AddEndpointsToSubnets" -> "AddMissingAndInvalidEndpointsToSubnets" or "AddMissingAndNotSuccededEndpointsToSubnet" ? 

"addEndpointsToSubnet" -> "addMissingAndInvalidEndpointsToSubnet" or "addMissingAndNotSuccededEndpointsToSubnet"

Any suggestion is welcome.

Tks!

Reply:
> Also I missed it when you updated it, but when is the pointer nil and endpointFound==true ?
Your pointer check is superfluous

What do you mean? I need to do 
```go
serviceEndpointPtr != nil
```
if not, 
```go
serviceEndpointPtr.ProvisioningState
````
on line 33 can panic, no? Maybe I am missing something.

I think I am not checking for "pointer nil and endpointFound==true".

Reply:
A documentation comment is fine, and that would show up in function hovering if it starts with the function name. 

Basically , the pointer is nil only if `subnetContains... ` returns false. But if it returns false then go will stop at the first condition (which is true because it is `!containsEndpoint`), and will never evaluate the second value of the `or` statement https://go.dev/ref/spec#Logical_operators

Reply:
> Basically , the pointer is nil only if subnetContains... returns false. But if it returns false then go will stop at the first condition (which is true because it is !containsEndpoint), and will never evaluate the second value of the or statement

> Your pointer check is superfluous

You are wrong about not checking the pointer against nil. 

Facts (check the attached picture): I removed the nil check that you mention and tests executing that part fail exactly at line 33 (with the nil check that already exists the tests don't fail).

Why?
For the simple fact that here (line 33 of pkg/util/subnet/subnetsupdater.go):
```go
endpointSucceded := serviceEndpointPtr != nil && serviceEndpointPtr.ProvisioningState == mgmtnetwork.Succeeded
```
at exactly that point, you are not evaluating the if condition that you mention, that will come later. You just thought about what happens on the execution of the "if" on line 35
but you completely forgot that the nil check is necessary because at line 33 nothing guarantees you that serviceEndpointPtr is not nil. No change is needed.

![tests_fail_without_checking_nil](https://user-images.githubusercontent.com/38758944/180304841-2f3cce9f-7f6c-479e-a813-3583eb07a56e.png)

Reply:
that's what you get for using a gazillion useless variables. it was working before when there was still a `continue` statement, so there is no reason it is not working now. 
please stop trying to reinvent the wheel.

it should be like this:
```go
func addEndpointsToSubnet(endpoints []string, subnet *mgmtnetwork.Subnet) (subnetChanged bool) {
	for _, endpoint := range endpoints {
		endpointFound, serviceEndpointPtr := subnetContainsEndpoint(subnet, endpoint)

		if !endpointFound || serviceEndpointPtr.ProvisioningState != mgmtnetwork.Succeeded {
			addEndpointToSubnet(endpoint, subnet)
			subnetChanged = true
		}
	}

	return subnetChanged
}
```

Reply:
> that's what you get for using a gazillion useless variables. 
No, you just proposed something that was not working either because you did not check the current status of the code at that point (your comment proposing a nonworking change was done 5 hours after the last push) or you made some assumptions that break the code as I explained in the previous comment.

"gazillion useless variables"-> this was just one explanatory variable to make the "if" easier to read. I will now remove the variable and just do an "or" as your last suggestion because now it does work and is okay for me.

Thanks! 🙂